### PR TITLE
ISPN-15047 ZRANGESTORE

### PR DESCRIPTION
**Pull request by karesti**
https://issues.redhat.com/browse/ISPN-15047

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 1]
```diff
@@ -213,6 +222,20 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, Channel
          builder.stop(to);
          getSortedSetCall = sortedSetCache.subsetByIndex(name, builder.build());
       }
+      if (destination != null) {
+         // Store range and return size
+         return CompletionStages.handleAndCompose(getSortedSetCall, (scoredValues, t1) -> {
+            CompletionStage<Long> sortedSetSize;
+            if (scoredValues.size() > 0) {
+               sortedSetSize = sortedSetCache.addMany(destination, applyLimit(scoredValues, resultOpt), SortedSetAddArgs.create().returnSize().build());
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:230]: indeed, I believed it was an add to an existing one

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 2]
```diff
@@ -226,6 +249,34 @@ private static CompletionStage<RespRequestHandler> returnEmptyList(Resp3Handler
...
    }
 
+   private Collection<SortedSetBucket.ScoredValue<byte[]>> applyLimit(Collection<SortedSetBucket.ScoredValue<byte[]>> scoredValues, ResultOptions resultOpt) {
+      if (!resultOpt.limit()) {
+         return scoredValues;
+      }
+
+      List<SortedSetBucket.ScoredValue<byte[]>> limitedElements = new ArrayList<>();
+      Iterator<SortedSetBucket.ScoredValue<byte[]>> ite = scoredValues.iterator();
+      if (resultOpt.limit()) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:259]: yep

File: server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java [hunk 3]
```diff
@@ -213,6 +222,20 @@ public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, Channel
...
          getSortedSetCall = sortedSetCache.subsetByIndex(name, builder.build());
       }
+      if (destination != null) {
+         // Store range and return size
+         return CompletionStages.handleAndCompose(getSortedSetCall, (scoredValues, t1) -> {
+            CompletionStage<Long> sortedSetSize;
+            if (scoredValues.size() > 0) {
+               sortedSetSize = sortedSetCache.addMany(destination, applyLimit(scoredValues, resultOpt), SortedSetAddArgs.create().returnSize().build());
+            } else {
+               sortedSetSize = CompletableFuture.completedFuture(0L);
```
[the remark this answers is not in the document]
Comment (karesti) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java:232]: thanks

**Review by jabolina — COMMENTED** [review]
Pretty good. Just the exception thing and we're good.

File: multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java [hunk 4]
```diff
@@ -291,10 +350,10 @@ public long size() {
 
    @ProtoTypeId(ProtoStreamTypeIds.MULTIMAP_SORTED_SET_SCORED_ENTRY)
    public static class ScoredValue<V> implements Comparable<ScoredValue<V>> {
-      private final double score;
+      private final Double score;
       private final MultimapObjectWrapper<V> value;
 
-      private ScoredValue(double score, V value) {
+      private ScoredValue(Double score, V value) {
```
[the remark this answers is not in the document]
Comment (karesti) [reply multimap/src/main/java/org/infinispan/multimap/impl/SortedSetBucket.java:356]: yes, this is possible in ZMSCORE. if the value-score does not exist the value holds null score. 
probably I added this one here related to my local work on other commands that are coming

**Review by jabolina — APPROVED** [review]

**Comment by jabolina** [comment]
Thanks, @karesti!

**Pull request merged**